### PR TITLE
feat(instrumentation): add streamed OpenAI function tracing

### PR DESCRIPTION
**Pull request by achang97**
- [x] I have added tests that cover my changes.
- [ ] If adding a new instrumentation or changing an existing one, I've added screenshots from some observability platform showing the change.
- [x] PR name follows conventional commits format: `feat(instrumentation): ...` or `fix(instrumentation): ...`.
- [ ] (If applicable) I have updated the documentation accordingly.

Fixes https://github.com/traceloop/openllmetry/issues/1283

### CODE REVIEW

File: packages/opentelemetry-instrumentation-openai/tests/traces/test_chat.py [hunk 1]
```diff
@@ -33,6 +56,44 @@ def test_chat(exporter, openai_client):
...
+        model="gpt-3.5-turbo",
+        messages=[
+            {"role": "user", "content": "What's the weather like in San Francisco?"}
+        ],
+        tools=openai_tools,
+    )
+
+    spans = exporter.get_finished_spans()
+
+    assert [span.name for span in spans] == [
```
Comment (achang97) [inline packages/opentelemetry-instrumentation-openai/tests/traces/test_chat.py:71]: These tests are redundant for the 3 cases (non-streaming, streaming, async streaming). Should I use a shared helper function here for the `assert` statements?

Comment (nirga) [reply packages/opentelemetry-instrumentation-openai/tests/traces/test_chat.py:71]: No, it's better to be explicit in tests, even if it's repetitive. Maybe we can switch to a parameterized test (but not important for this PR)

**Review by nirga — APPROVED** [review]

File: packages/opentelemetry-instrumentation-openai/opentelemetry/instrumentation/openai/shared/chat_wrappers.py [hunk 2]
```diff
@@ -551,6 +551,18 @@ def _process_item(self, item):
...
             if delta and delta.get("role"):
                 complete_choice["message"]["role"] = delta.get("role")
+            if delta and delta.get("tool_calls"):
+                tool_call = delta.get("tool_calls")[0]
+                if "tool_calls" not in complete_choice["message"]:
+                    complete_choice["message"]["tool_calls"] = [
+                        {"function": {"name": "", "arguments": ""}}
+                    ]
+
+                function = complete_choice["message"]["tool_calls"][0]["function"]
```
[the remark this answers is not in the document]
Comment (achang97) [reply packages/opentelemetry-instrumentation-openai/opentelemetry/instrumentation/openai/shared/chat_wrappers.py:561]: I see -- do you mean using the `.get` syntax instead?

[the remark this answers is not in the document]
Comment (nirga) [reply packages/opentelemetry-instrumentation-openai/opentelemetry/instrumentation/openai/shared/chat_wrappers.py:561]: Yes, but also testing that there's a tool, and index 0 exists

[the remark this answers is not in the document]
Comment (achang97) [reply packages/opentelemetry-instrumentation-openai/opentelemetry/instrumentation/openai/shared/chat_wrappers.py:561]: Got it, do you mean that I need to add to the existing tests? Can add a explicit length check before accessing index 0

**Review by nirga — APPROVED** [review]

**Pull request merged**